engine/access/rpc/backend/backend.go

```diff
@@ -226,12 +230,80 @@ func (b *Backend) GetNetworkParameters(_ context.Context) access.NetworkParamete
 
 // GetLatestProtocolStateSnapshot returns the latest finalized snapshot
 func (b *Backend) GetLatestProtocolStateSnapshot(_ context.Context) ([]byte, error) {
-	data, err := convert.SnapshotToBytes(b.state.Final())
+	snapshot := b.state.Final()
+
+	validSnapshot, err := b.getValidSnapshot(snapshot, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return convert.SnapshotToBytes(validSnapshot)
+}
+
+// getValidSnapshot will return a valid snapshot that has a sealing segment which
+// 1. does not contain any blocks that span an epoch transition
+// 2. does not contain any blocks that span an epoch phase transition
+// If a snapshot does contain an invalid sealing segment query the state
+// by height of each block in the segment and return a snapshot at the point
+// where the transition happens.
+func (b *Backend) getValidSnapshot(snapshot protocol.Snapshot, blocksVisited int) (protocol.Snapshot, error) {
+	if blocksVisited > snapshotHistoryLimit {
+		return nil, fmt.Errorf("failed to get valid snapshot reached look back limit ")
+	}
+
+	segment, err := snapshot.SealingSegment()
+	if err != nil {
+		return nil, err
+	}
+
+	// get snapshots at the highest and lowest blocks in segment for comparison
+	snapshotAtHighest := b.state.AtHeight(segment.Highest().Header.Height)
+	snapshotAtLowest := b.state.AtHeight(segment.Lowest().Header.Height)
+
+	counterAtHighest, err := snapshotAtHighest.Epochs().Current().Counter()
+	if err != nil {
+		return nil, err
+	}
+
+	phaseAtHighest, err := snapshotAtHighest.Phase()
 	if err != nil {
 		return nil, err
 	}
 
-	return data, nil
+	counterAtLowest, err := snapshotAtLowest.Epochs().Current().Counter()
+	if err != nil {
+		return nil, err
+	}
+
+	phaseAtLowest, err := snapshotAtLowest.Phase()
+	if err != nil {
+		return nil, err
+	}
+
+	// check if any epoch or phase transition occurs, visit each node
+	// in strict order of decreasing height starting at head
+	if counterAtHighest != counterAtLowest || phaseAtHighest != phaseAtLowest {
+		for i := len(segment.Blocks) - 1; i >= 0; i-- {
```

Comment:
I don't think we need query the segment for every height. Instead we could do:
```
if a segment has crossed epoch
  if no, then return the snapshot
  if yes, then iterate through heights until finding the first height that has different phase than the head's phase
    recur
```

The logic would be like this
```
func getValidSnapshot(snapshot) {
  segment := snapshot.Segment()
  lowestEpoch, lowestPhase, lowestHeight := getLowest(segment)
  highestEpoch, highestPhase, highestHeight := getHighest(segment)
  if lowestEpoch == highestEpoch && lowestPhase == highestPhase {
    return snapshot
  }

  // find the first height that is in a different epoch/phase
  loop from highest to lowest {
    epoch, phase := getEpochPhase(height)
    if epoch == highestEpoch && phase == highestPhase {
      continue
    }

    // recur again to check the segment no longer cross epoch/phase
    //  if still crossing, then it will get the snapshot of previous 
    return getValidSnapshot(getSnapshotAtHeight(height))
  } 
}

```

Also I wonder if possible we can get the height range for epoch or phase if so we can further simplify to be:
```
func getValidSnapshot(snapshot) {
  segment := snapshot.Segment()
  lowestEpoch, lowestPhase, lowestHeight := getLowest(segment)
  highestEpoch, highestPhase, highestHeight := getHighest(segment)
  if lowestEpoch == highestEpoch && lowestPhase == highestPhase {
    return snapshot
  }

  epoch, phase := getPreviousEpochPhase(lowestEpoch, lowestPhase)
  startHeight, endHeight := getHeightRange(epoch, phase)
  // recur again to check the segment no longer cross epoch/phase
  //  if still crossing, then it will get the snapshot of previous 

  return getValidSnapshot(getSnapshotAtHeight(endHeight))
}
```

Reply:
How does this differ from the current implementation aside from moving the logic to get phase and counter at a specific block into helper funcs **_getLowest(segment)_** _**getHighest(segment)**_ ? Currently we check if there is a epoch or phase transition (in the unhappy path we will recur until we get a valid snapshot ) , if not we simply return the snapshot requested.

Reply:
Also should note we are not querying the state for every height, current implementation checks if there is a difference in epoch or epoch phase at each height in the segment starting at the highest. In most sealing segments (5-10) blocks this check will only be about 3-4 blocks before finding the block on the boundary of the transition. Only in the worst case scenario (segment with a lot of blocks) will we iterate deep into the segment blocks.

Reply:
Let's take an example:
Epoch1's three phases are ranging from height: `[1001-1500][1501-1800][1801-1840]`
Epoch2's three phases are ranging from height: `[1841-2500][2501-3000][3001-3080]`

Now, if `GetLatestProtocolStateSnapshot` is called at block 2530, which is at the 2nd phase of Epoch2, then you only do the following querys:
1. find the segment, and it returns the block of the segment ranges from `[2460-2530]`, which goes from Epoch2's 1st phase to Epoch2's 2nd phase.
2. Instead of going 1 height at a time from `2530,2529,2528,...2501,2500` until reaching 2500 to find the first block that is in a different phase, which will loop 30 times, we could jump to 2500 right away, because we know it has crossed the phase, and we the previous phase is phase 1, and it's block range is `[1841-2500]` so I can get the snapshot of the end height (2500) of previous phase, and recursively run the check again.

If we want to keep track of the blockVisited, we could add the range that we jumped: 
```
func getValidSnapshot(snapshot, blockVisited) (snapshot, error) {
  if blockVisited > limit {
    return nil, err
  }

  segment := snapshot.Segment()
  lowestEpoch, lowestPhase, lowestHeight := getLowest(segment)
  highestEpoch, highestPhase, highestHeight := getHighest(segment)
  if lowestEpoch == highestEpoch && lowestPhase == highestPhase {
    return snapshot
  }

  epoch, phase := getPreviousEpochPhase(lowestEpoch, lowestPhase)
  startHeight, endHeight := getHeightRange(epoch, phase)
  // recur again to check the segment no longer cross epoch/phase
  //  if still crossing, then it will get the snapshot of previous 

  return getValidSnapshot(getSnapshotAtHeight(endHeight), blockVisited + (highestHeight - endHeight))
}
```

Reply:
Correct me if i'm wrong but at the moment the endpoint has no information about the height ranges of phases or when a phase starts or stops. We only know the current phase we are in for the given snapshot, this is not possible without that information.

Reply:
Unfortunately we don't index the heights of epoch/phase transitions, so we do need to search the segment to find the transition point. 

In general, I don't think we need to over-engineer this too much for a few reasons:
* The vast majority of sealing segments do not transit epoch/phase transitions (we will very, very rarely enter the linear search)
* The vast majority of sealing segments are <10 blocks (when we do enter the linear search, it is inexpensive)
* In practice, if a sealing segment transits an epoch/phase transition, the next segment we check will not, and we will be able to return without a second linear search
* If any of the above points are not true, then we want to fail fast without returning any snapshot at all, because it will be either a huge amount of data, or very outdated (ie. we are explicitly not trying to handle the expensive edge cases of this function)

Reply:
If we can't get the height range for a given epoch/phase, then what do you think about my first suggestion?

```
func getValidSnapshot(snapshot) {
  segment := snapshot.Segment()
  lowestEpoch, lowestPhase, lowestHeight := getLowest(segment)
  highestEpoch, highestPhase, highestHeight := getHighest(segment)
  if lowestEpoch == highestEpoch && lowestPhase == highestPhase {
    return snapshot
  }

  // find the first height that is in a different epoch/phase
  loop from highest to lowest {
    epoch, phase := getEpochPhase(height)
    if epoch == highestEpoch && phase == highestPhase {
      continue
    }

    // recur again to check the segment no longer cross epoch/phase
    //  if still crossing, then it will get the snapshot of previous 
    return getValidSnapshot(getSnapshotAtHeight(height))
  } 
}
```
What this improves is that, we don't need to query `Segment()` when iterate through each block, because iterating through blocks is just to find the first block that's in a different phase.

Reply:
> what do you think about my first suggestion?

I think it's good but it seems functionally equivalent to the current implementation. (Though it is clearer to read, so for that reason I think it's worth adopting anyway.)

For example, instead of:
```
if lowestEpoch == highestEpoch && lowestPhase == highestPhase {
    return snapshot
}
```
we have the inverse conditional for entering the segment iteration:
```
if counterAtHighest != counterAtLowest || phaseAtHighest != phaseAtLowest { ... }
```

> What this improves is that, we don't need to query Segment() when iterate through each block

I don't see where this is happening currently 😅

Reply:
We can discuss after PE standup, I still don't see any difference in either of your suggestions from what's currently implemented. Maybe another set of eyes would help cc: @jordanschalm

Reply:
I see. Reading through the logic again makes sense now. I was a bit confused by this check
```
counterAtHighest != counterAtLowest || phaseAtHighest != phaseAtLowest
```

Maybe we could extract into a function:
```
func isEpochOrPhaseDifferent(counter1, counter2, phase1, phase2) bool
```

Reply:
@jordanschalm : _we have the inverse conditional for entering the segment iteration:_

This is because of the semantics of the recursion. In @zhangchiqing suggestion there is no return statement at the end of the func . Thus it's actually much clearer to read currently. 

- get counter and phases 
- check them 
  - if transition, check for block that straddles boundary and do recursion
 - return